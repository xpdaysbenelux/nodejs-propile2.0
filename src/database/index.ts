export * from './entities';
export * from './repositories';
export * from './database.module';
