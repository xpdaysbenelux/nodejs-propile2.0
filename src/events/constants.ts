export enum EventTitle {
    Break = 'BREAK',
    Closing = 'CLOSING',
    CoffeeBreak = 'COFFEE_BREAK',
    ConferenceDinner = 'CONFERENCE_DINNER',
    Dinner = 'DINNER',
    Drinks = 'DRINKS',
    EveningProgramme = 'EVENING_PROGRAMME',
    Lunch = 'LUNCH',
    Plenary = 'PLENARY',
    Registration = 'REGISTRATION',
    RegistrationAndCoffee = 'REGISTRATION_AND_COFFEE',
    Welcome = 'WELCOME',
}
