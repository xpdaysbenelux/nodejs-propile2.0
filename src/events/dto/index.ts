export * from './create-event.dto';
export * from './get-event.dto';
export * from './update-event.dto';
