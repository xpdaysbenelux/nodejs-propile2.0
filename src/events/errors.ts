import { BadRequestException, NotFoundException } from '@nestjs/common';

export class EndTimeMustBeLaterThanStartTime extends BadRequestException {
    constructor() {
        super(
            'The end time of an event must be later than the start time!',
            'END_TIME_LATER_THEN_START_TIME',
        );
    }
}

export class EventNotFound extends NotFoundException {
    constructor() {
        super('The event was not found.', 'EVENT_NOT_FOUND');
    }
}
