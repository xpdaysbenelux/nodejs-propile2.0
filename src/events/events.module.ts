import { Module } from '@nestjs/common';

import { EventController } from './events.controller';
import { EventsService } from './events.service';
import { EventsQueries } from './events.queries';
import { DatabaseModule } from '../database';

@Module({
    imports: [DatabaseModule],
    controllers: [EventController],
    providers: [EventsService, EventsQueries],
})
export class EventsModule {}
