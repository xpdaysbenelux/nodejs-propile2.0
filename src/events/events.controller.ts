import {
    Body,
    UseGuards,
    Controller,
    Post,
    Get,
    Param,
    Put,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { EventsService } from './events.service';
import { EventsQueries } from './events.queries';
import { UserSession, RequiredPermissions } from '../_shared/decorators';
import { IUserSession } from '../_shared/constants';
import {
    AuthenticatedGuard,
    RequiredPermissionsGuard,
} from '../_shared/guards';
import { CreateEventRequest, EventResponse, EventIdParam } from './dto';
import { ProgramIdParam } from '../programs/dto';

@UseGuards(AuthenticatedGuard)
@Controller('/programs/:programId/events')
@ApiTags('events')
export class EventController {
    constructor(
        private readonly eventsService: EventsService,
        private readonly eventsQueries: EventsQueries,
    ) {}

    @RequiredPermissions({ programs: { view: true, edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get(':eventId')
    getEvent(@Param() params: EventIdParam): Promise<EventResponse> {
        return this.eventsQueries.getEvent(params.eventId);
    }

    @RequiredPermissions({ programs: { view: true, edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get()
    getEvents(@Param() params: ProgramIdParam): Promise<EventResponse[]> {
        console.log('events', params.programId);
        return this.eventsQueries.getEvents(params.programId);
    }

    @RequiredPermissions({ programs: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Post()
    async createEvent(
        @Body() body: CreateEventRequest,
        @UserSession() session: IUserSession,
    ): Promise<EventResponse> {
        const eventId = await this.eventsService.createEvent(body, session);
        return this.eventsQueries.getEvent(eventId);
    }

    @RequiredPermissions({ programs: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Put(':eventId')
    async updateEvent(
        @Body() body: CreateEventRequest,
        @Param() params: EventIdParam,
        @UserSession() session: IUserSession,
    ): Promise<EventResponse> {
        const eventId = await this.eventsService.updateEvent(
            body,
            params.eventId,
            session,
        );
        return this.eventsQueries.getEvent(eventId);
    }
}
