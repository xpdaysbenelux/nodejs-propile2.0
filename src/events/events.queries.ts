import { Injectable } from '@nestjs/common';

import { EventRepository } from '../database';
import { EventResponse } from './dto';

const eventFields = [
    'event.id',
    'event.title',
    'event.spanRow',
    'event.comment',
    'event.startTime',
    'event.endTime',
    'program.id',
    'program.title',
    'program.date',
    'program.startTime',
    'program.endTime',
    'room.id',
    'room.name',
    'room.maxParticipants',
    'session.id',
    'session.title',
    'session.duration',
    'session.firstPresenter',
    'session.secondPresenter',
    'session.maxParticipants',
];

@Injectable()
export class EventsQueries {
    constructor(private readonly eventRepository: EventRepository) {}

    async getEvent(eventId: string): Promise<EventResponse> {
        return this.eventRepository
            .createQueryBuilder('event')
            .select(eventFields)
            .where('event.id = :eventId', { eventId })
            .innerJoin('event.program', 'program')
            .leftJoin('event.room', 'room')
            .leftJoin('event.session', 'session')
            .getOne();
    }
}
