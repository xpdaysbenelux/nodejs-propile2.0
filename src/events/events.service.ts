import { Injectable } from '@nestjs/common';
import { parseISO, isAfter } from 'date-fns';

import {
    EventRepository,
    Event,
    ProgramRepository,
    RoomRepository,
    SessionRepository,
} from '../database';
import { CreateEventRequest } from './dto/create-event.dto';
import { IUserSession } from '../_shared/constants';
import { ProgramNotFoud } from '../programs/errors';
import { RoomNotFound } from '../rooms/errors';
import { SessionNotFound } from '../sessions/errors';
import { EndTimeMustBeLaterThanStartTime, EventNotFound } from './errors';
import { UpdateEventRequest } from './dto';

@Injectable()
export class EventsService {
    constructor(
        private readonly eventRepository: EventRepository,
        private readonly programRepository: ProgramRepository,
        private readonly roomRepository: RoomRepository,
        private readonly sessionRepository: SessionRepository,
    ) {}

    async createEvent(
        body: CreateEventRequest,
        user: IUserSession,
    ): Promise<string> {
        const {
            spanRow,
            programId,
            title,
            startTime,
            endTime,
            comment,
            roomId,
            sessionId,
        } = body;
        const eventStartTime = parseISO(startTime);
        const eventEndTime = parseISO(endTime);

        const event = new Event();

        const [program, room, session] = await Promise.all([
            this.programRepository.findOne({
                id: programId,
            }),
            this.roomRepository.findOne({
                id: roomId,
            }),
            this.sessionRepository.findOne({
                id: sessionId,
            }),
        ]);

        if (!program) {
            throw new ProgramNotFoud();
        }
        event.program = program;

        if (roomId) {
            if (!room) {
                throw new RoomNotFound();
            }
            event.room = room;
        } else event.room = null;

        if (sessionId) {
            if (!session) {
                throw new SessionNotFound();
            }
            event.session = session;
        } else event.session = null;

        if (title) event.title = title;
        else event.title = null;

        if (!isAfter(eventEndTime, eventStartTime)) {
            throw new EndTimeMustBeLaterThanStartTime();
        }
        event.startTime = eventStartTime;
        event.endTime = eventEndTime;
        event.spanRow = spanRow;
        event.comment = comment;
        event.createdBy = user.email;
        event.createdAt = new Date();

        await this.eventRepository.save(event);
        return event.id;
    }

    async updateEvent(
        body: UpdateEventRequest,
        eventId: string,
        user: IUserSession,
    ): Promise<string> {
        const {
            spanRow,
            programId,
            title,
            startTime,
            endTime,
            comment,
            roomId,
            sessionId,
        } = body;
        const eventStartTime = parseISO(startTime);
        const eventEndTime = parseISO(endTime);

        const [existingEvent, program, room, session] = await Promise.all([
            this.eventRepository.findOne({ id: eventId }),
            this.programRepository.findOne({
                id: programId,
            }),
            this.roomRepository.findOne({
                id: roomId,
            }),
            this.sessionRepository.findOne({
                id: sessionId,
            }),
        ]);

        if (!existingEvent) throw new EventNotFound();

        if (!program) {
            throw new ProgramNotFoud();
        }
        existingEvent.program = program;

        if (roomId) {
            if (!room) {
                throw new RoomNotFound();
            }
            existingEvent.room = room;
        } else existingEvent.room = null;

        if (sessionId) {
            if (!session) {
                throw new SessionNotFound();
            }
            existingEvent.session = session;
        } else existingEvent.session = null;

        if (title) existingEvent.title = title;
        else existingEvent.title = null;

        if (!isAfter(eventEndTime, eventStartTime)) {
            throw new EndTimeMustBeLaterThanStartTime();
        }
        existingEvent.startTime = eventStartTime;
        existingEvent.endTime = eventEndTime;
        existingEvent.spanRow = spanRow;
        existingEvent.comment = comment;
        existingEvent.createdBy = user.email;
        existingEvent.createdAt = new Date();

        await this.eventRepository.save(existingEvent);
        return existingEvent.id;
    }

    async deleteEvent(eventId: string): Promise<string> {
        const existingEvent = await this.eventRepository.findOne({
            where: { id: eventId },
        });
        if (!existingEvent) {
            throw new EventNotFound();
        }

        await this.eventRepository.delete(eventId);
        return eventId;
    }
}
