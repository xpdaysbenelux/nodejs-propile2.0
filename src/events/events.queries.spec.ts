import { Test } from '@nestjs/testing';
import * as faker from 'faker';

import { EventRepository, DatabaseModule } from '../database';
import { EventsQueries } from './events.queries';

describe('EventsQueries', () => {
    let eventRepository: EventRepository;
    let eventsQueries: EventsQueries;

    beforeAll(async () => {
        const module = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [EventsQueries],
        }).compile();

        eventsQueries = module.get(EventsQueries);
        eventRepository = module.get(EventRepository);
    });

    afterAll(() => {
        eventRepository.manager.connection.close();
    });

    describe('getEvent', () => {
        it('should return the requested event correctly', async () => {
            const result = await eventsQueries.getEvent(
                '6ee1e9e0-1b79-49dd-b326-09e992430c31',
            );

            const { startTime, endTime, program, ...withoutDates } = result;
            expect(withoutDates).toMatchSnapshot();
        });

        it('should return nothing if the requested event does not exist', async () => {
            const result = await eventsQueries.getEvent(faker.random.uuid());

            expect(result).toMatchSnapshot();
        });
    });

    describe('getEvents', () => {
        it('should return the requested events by program id', async () => {
            const result = await eventsQueries.getEvents(
                '3185e221-73ca-4b5a-93a9-1cc2d0b5df76',
            );

            const withoutDatesList = result.map(entry => {
                const { startTime, endTime, program, ...withoutDates } = entry;
                return withoutDates;
            });

            expect(withoutDatesList).toMatchSnapshot();
        });

        it('should return nothing if there are no events found for the given program id', async () => {
            const result = await eventsQueries.getEvents(faker.random.uuid());

            expect(result).toMatchSnapshot();
        });
    });
});
