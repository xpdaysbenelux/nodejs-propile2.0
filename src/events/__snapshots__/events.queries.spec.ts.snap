// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsQueries getEvent should return nothing if the requested event does not exist 1`] = `undefined`;

exports[`EventsQueries getEvent should return the requested event correctly 1`] = `
Object {
  "comment": "Lorem ipsum dolor sit amet",
  "id": "6ee1e9e0-1b79-49dd-b326-09e992430c31",
  "room": null,
  "session": null,
  "spanRow": true,
  "title": "LUNCH",
}
`;

exports[`EventsQueries getEvents should return nothing if there are no events found for the given program id 1`] = `Array []`;

exports[`EventsQueries getEvents should return the requested events by program id 1`] = `
Array [
  Object {
    "comment": null,
    "id": "4a63356b-3ca6-40c6-84ff-2b30360baf08",
    "room": Room {
      "id": "ce63d11d-c51f-4320-a997-fb31b42a6e08",
      "maxParticipants": 50,
      "name": "Room 2",
    },
    "session": Session {
      "duration": "30",
      "id": "8c87028d-31f6-414b-b02d-e7431b6a7243",
      "maxParticipants": 25,
      "title": "How to...",
    },
    "spanRow": false,
    "title": null,
  },
  Object {
    "comment": "Lorem ipsum dolor sit amet",
    "id": "6ee1e9e0-1b79-49dd-b326-09e992430c31",
    "room": null,
    "session": null,
    "spanRow": true,
    "title": "LUNCH",
  },
]
`;
